src/main/java/org/hydev/mcpm/client/interaction/ProgressRow.java

```diff
@@ -0,0 +1,134 @@
+package org.hydev.mcpm.client.interaction;
+
+import static java.lang.String.format;
+
+/**
+ * Row of a progress bar
+ *
+ * @author Azalea (https://github.com/hykilpikonna)
+ * @since 2022-10-30
+ */
+public class ProgressRow
+{
+    private final long total;
+    private long completed;
+    private String unit;
+    private ProgressBar pb;
+    private String desc;
+    private int descLen;
+    private String fmt;
+
+    private final long startTime;
+
+    public ProgressRow(long total)
+    {
+        this.total = total;
+        this.completed = 0;
+        this.unit = " it";
+        this.desc = "";
+        this.descLen = 20;
+        this.fmt = "{desc}{speed} {eta} {prefix}{progbar}{suffix} {%done}";
+
+        // Record start time for speed estimation
+        this.startTime = System.nanoTime();
+    }
+
+    /**
+     * @return Elapsed time in seconds
+     */
+    private double elapsed()
+    {
+        return (System.nanoTime() - startTime) / 1e9d;
+    }
+
+    /**
+     * Get formatted string of the current progress bar
+     *
+     * @param theme Progress bar theme
+     * @param cols Number of columns (width) of the terminal window
+     * @return Formatted string
+     */
+    public String toString(ProgressBarTheme theme, int cols)
+    {
+        // Calculate speed. TODO: Use a moving window to calculate speed
+        double speed = completed / elapsed();
+        double eta = total / speed;
+        long etaS = (long) (eta % 60), etaM = (long) (eta / 60);
+
+        // Replace variables
+        var p = format("%3.0f%%", 100d * completed / total);
+        var t = fmt.replace("{prefix}", theme.prefix())
```

Comment:
Not trying to say we should pull in a whole templating library just for this, but can we do anything to sidestep String#replace? Weird edge cases include theme.suffix() returning something like `"{eta}"` or `desc` being `"{progbar} {progbar} {progbar}"`.

Reply:
Yea but we can just make sure that doesn't happen since we're only using it internally in our program, and it's really unlikely to happen just by chance...

Reply:
What about file names? Those are user controlled so someone who knows about our utility can break this class. Again, unlikely, but it's spooky to give the user control over our format strings.

Reply:
This is why description is replaced at the very last, so that the placeholders in description wouldn't be replaced by anything else